docs/proposals/affinity.md

```diff
@@ -0,0 +1,411 @@
+whether "put" means "must put" (hard) or "try to put" (soft)--all that changes is whether
+the rule goes in HardPodAffinity or SoftPodAffinity. Also, some of the examples
+use "zone" and some use "node," just to make the examples more interesting; any of the examples
+with "zone" will also work for "node" if you change the `TopologyKey`, and vice-versa.
+
+* **Put the pod in zone Z**:
+Tricked you! It is not possible express this using the API described here. For this you should use a HardNodeAffinity.
+
+* **Put the pod in a zone that is running at least one pod from service S**:
+`{LabelSelector: <selector that matches S's pods>, TopologyKey: "zone"}`
+
+* **Put the pod on a node that is already running a pod that requires a license for software package P**:
+Assuming pods that require a license for software package P have a label `{key=license, value=P}`:
+`{LabelSelector: "license" In "P", TopologyKey: "node"}`
+
+* **Put this pod in the same zone as other pods from its same service**:
+Assuming pods from this pod's service have some label `{key=service, value=S}`:
+`{LabelSelector: "service" In "S", TopologyKey: "zone"}`
+
+This last example illustrates a small issue with this API when it is used
+with a scheduler that processes the pending queue one pod at a time, like the current
+Kubernetes scheduler. The HardAffinityExpression
+`{LabelSelector: "service" In "S", TopologyKey: "zone"}`
+only "works" once one pod from service S has been scheduled. But if all pods in service
+S have this HardAffinityExpression in their PodSpec, then the HardAffinityExpression will block the first
+pod of the service from ever scheduling, since it is only allowed to run in a zone with another pod from
+the same service. And of course that means none of the pods of the service will be able
+to schedule. This problem *only* applies to hard affinity, not soft affinity or hard or soft
+anti-affinity. There are at least two ways to solve this problem
+* short-term: when a controller creates pods from a PodTemplate, for exactly *one* of those
+pods, it should omit any hard affinity rules that select the pods of that PodTemplate.
+* long-term/speculative: controllers could present the scheduler with a group of pods from
+the same PodTemplate as a single unit. This would allow two things: (1) the scheduler could ignore
+problematic hard affinity rules rather than requiring the controller to handle it,
+and (2) the scheduler could do proper [gang scheduling](https://github.com/kubernetes/kubernetes/issues/16845)
+since it could receive an entire gang simultaneously.
+
+### Anti-affinity
+
+* **Spread the pods of this service S across nodes and zones**:
+`{{LabelSelector: <selector that matches S's pods>, TopologyKey: "node"}, {LabelSelector: <selector that matches S's pods>, TopologyKey: "zone"}}`
+
+* **Don't co-locate pods of this service with pods from service "evilService"**:
+`{LabelSelector: selector that matches evilService's pods, TopologyKey: "node"}`
```

Comment:
Got it.  I have no strong issue with the example, other than that it might encourage people to try to do this.  And that path leads to madness: it doesn't make sense to list all the possible "other services".

A bigger issue is whether we want to support anti-affinity at all - at least to start with. it's a powerful tool, but feels a bit like a knife without a handle.  It may be that we need it for spreading for failure tolerance, but perhaps there's a better way to achieve that narrow goal?